src/holder.rs

```diff
@@ -1,7 +1,79 @@
 use crate::sync::atomic::AtomicPtr;
 use crate::{Domain, HazPtrObject, HazPtrRecord};
+use std::mem::{ManuallyDrop, MaybeUninit};
 use std::sync::atomic::Ordering;
 
+pub struct HazardPointerArray<'domain, F, const N: usize> {
+    haz_ptrs: [ManuallyDrop<HazardPointer<'domain, F>>; N],
+}
+
+impl<'domain, F, const N: usize> HazardPointerArray<'domain, F, N> {
+    pub fn hazard_pointers<'array>(&'array mut self) -> [&'array mut HazardPointer<'domain, F>; N] {
+        let mut out: [MaybeUninit<&'array mut HazardPointer<'domain, F>>; N] =
+            [(); N].map(|_| MaybeUninit::uninit());
+
+        for (i, hazptr) in self.haz_ptrs.iter_mut().enumerate() {
+            out[i].write(hazptr);
+        }
+
+        //
+        // # Safety
+        //
+        // We have initialized every element of the array with our for loop above
+        out.map(|maybe_uninit| unsafe { maybe_uninit.assume_init() })
+    }
+
+    ///
+    /// # Safety
+    ///
+    /// Caller must guarantee that the address in `AtomicPtr` is valid as a reference, or null.
+    /// Caller must also guarantee that the value behind the `AtomicPtr` will only be deallocated
+    /// through calls to [`HazPtrObject::retire`] on the same [`Domain`] as this holder is
+    /// associated with.
+    pub unsafe fn protect_all<'l, 'o, T>(
+        &'l mut self,
+        mut sources: [&'_ AtomicPtr<T>; N],
+    ) -> [Option<&'l T>; N]
+    where
+        T: HazPtrObject<'o, F>,
+        'o: 'l,
+        F: 'static,
+    {
+        let mut out = [None; N];
+
+        for (i, (hazptr, src)) in self.haz_ptrs.iter_mut().zip(&mut sources).enumerate() {
+            out[i] = unsafe { hazptr.protect(src) };
+        }
+
+        out
+    }
+
+    pub fn reset_protection(&mut self) {
+        for hazptr in self.haz_ptrs.iter_mut() {
+            hazptr.reset_protection();
+        }
+    }
+}
+
+impl<'domain, F, const N: usize> Drop for HazardPointerArray<'domain, F, N> {
+    fn drop(&mut self) {
+        self.reset_protection();
+        let mut records: [MaybeUninit<&'domain HazPtrRecord>; N] = [MaybeUninit::uninit(); N];
+
+        for (i, haz_ptr) in self.haz_ptrs.iter_mut().enumerate() {
+            records[i].write(haz_ptr.hazard);
+        }
+
+        //
+        // # Safety
+        //
+        // We have initialized every element of the array with our for loop above
+        let records = records.map(|maybe_uninit| unsafe { maybe_uninit.assume_init() });
```

Comment:
Since we've used this pattern a few times now, let's write a helper function to do the `[HazPtrRecord; N]` -> `[&mut HazPtrRecord; N]` conversion.